### PR TITLE
Fixes installing lokal plugins with extras_require

### LINKED ISSUE #865: Wrong way to install dependencies of plugins
Hi there,

I've got the problem, that I can't install a local version of the Lektor-atom plugin, as it has an `extras_require`of `test`.

This leads to a `egg.info/requires.txt` of 

```
MarkupSafe
feedgenerator

[test]
lxml
pytest
```

Which of course is no valid requirements.txt file that pip can use, which then leads to an explosion when this is installed [here](https://github.com/lektor/lektor/blob/7132afbeaa64429dd1367dcb719573792a8b2f43/lektor/packages.py#L150).

I am not entirely sure what the correct way to consume this information is, but I suspect that one should go about using the `pkg_resources` package. Quick googling has provided [this stackoverflow](https://stackoverflow.com/a/50193639/4572750) which looks promising.

### PR DESCRIPTION
**Pull request by dwt**
The problem this tries to solve is that local plugin packages with `extras_require` fail to install. I had this problem with [lektor-atom](https://github.com/lektor/lektor-atom/) where I actually stumbled upon this issue here https://github.com/lektor/lektor-atom/issues/32

### Issue(s) Resolved

Fixes #865 

### Description of Changes
* [x] Wrote at least one-line docstrings (for any new functions)
* [x] Added unit test(s) covering the changes (if testable)

The pull request converts the requires.txt file from the egg_info call into a requirements.txt file, by stripping out everything from the first `[extras_require]` section in that file. It also documents why all of that is neccessary, as it would seem a much more logical choice to just call `pip install --target $folder --editable $plugin` and be done with it - but sadly that doesn't work because pip has a bug https://github.com/pypa/pip/issues/4390

### CODE REVIEW

File: lektor/packages.py
Comment (webknjaz) [inline lektor/packages.py]:
```suggestion
    # Because of these bugs:
```

File: lektor/packages.py
Comment (webknjaz) [inline lektor/packages.py]:
```suggestion
    """Create a sanitized copy of `requires.txt`."""
    # requires.txt can contain [extras_require] sections wich pip doesn't understand
```

File: tests/test_packages.py
Comment (webknjaz) [inline tests/test_packages.py]: Pytest already has a `tmpdir` fixture built-in. No need to reinvent it. That fixture is considered deprecated and you should use `tmp_path` instead (which returns `pathlib.Path()` objects).

Comment (dwt) [reply tests/test_packages.py]: Oh, I didn't know that existed - thanks for the hint.

**Comment by dwt** [comment]
@webknjaz That should be all of the points you brought up - I would be happy for a second round of review.

File: lektor/packages.py
Comment (dairiki) [inline lektor/packages.py]: Spelling aside, I wonder whether this long function name is really helping readability.

(The "in the same directory", I think, is an implementation detail and can be omitted from the function name with no loss of value.  The caller of this function really has no reason to care where the requirements.txt file is created.)

The essence of what this function does is: extract the `install_requires` from the egg-info `requires.txt`, and write them to `requirements.txt` suitable for passing to pip.

Note that we are lucky — and have been taking advantage of the fact — that the format of setuptool's `requires.txt` is similar to that of pip’s `requirements.txt`.  Really, their formats are defined by two different specifications for two different (though related) purposes.  `Requires.txt` is part of the egg-info metadata — its format is specified by setuptools (which is who defined the egg format). `Requirements.txt` is really mostly a way of passing arguments to pip — its format is specified by pip.

The reason the `requires.txt` was used as a `requirements.txt` in the first place was that is was easy.  Now that it's not quite as easy, maybe the use of a `requirements.txt` is not needed at all — just pass the requirements we'd like installed to pip as command-line arguments.

It might be cleaner to refactor a bit higher into `install_local_package` to make it clearer what is being done.  Perhaps:

```python
def install_local_package(package_root, path):
    """This installs a local dependency of a package."""
    # XXX: windows
    env = dict(os.environ)
    env["PYTHONPATH"] = package_root

    # Step 1: generate egg info and link us into the target folder.
    [...] (unchanged)

    # Step 2: install additional requirements
    for requirement in get_package_install_requires(path):
        # NB: this bastardizes the `package` argument of `download_and_install_package`
        # as we're passing a *requirement specifier* rather than just a package name.
        # I think it will work, but might want to adjust the signature of `download_and_install_package`
        # to provide a way to make this more explicit.  (While at it, perhaps provide a way to provide
        # multiple requirement specifiers in one go?)
        download_and_install_package(package_root, requirement)
```

The `get_package_install_requires` (probably) runs `setup.py egg_info` (as was done previously by `install_local_package`) and returns a sequence containing the initial lines — up to the first section marker ­of the resulting `requires.txt`.

Okay, I've rambled now, but, I think, have arrived at a worthwhile place:

The only reason `setup.py egg_info` is being run in `install_local_package` is to get at the packages `install_requires`.  Abstracting that whole process to a new function makes that logic clear.  Also, it abstracts implementation details — `install_local_package` doesn't care how the `install_requires` are obtained.

**Comment by yagebu** [comment]
I understand that limitations of `pip` force us to do some of this by hand, but wouldn't it be better to use `pkg_resources` to read the egg-info? This rewriting between two different formats specified by two different tools seems rather fragile.

**Comment by webknjaz** [comment]
> but wouldn't it be better to use `pkg_resources`

Even better: use `importlib_metadata`

**Comment by dairiki** [comment]
@webknjaz I was unaware of the emergence of `importlib.metadata`.  Good to know!

There may well be a way to make them work — if so, that would be the way to go — however, I just played a bit with both `pkg_resources` and `importlib.metadata` and could not get either of them to work with a "half-installed" (--no-deps) package, like we are working with here.  (When run for a --no-deps installed package with missing dependencies, `pkg_resources.require(<pkg>)` raises `DistributionNotFound`, `metadata(<pkg>)` raises 'PackageNotFoundError`.)

(The original author of this code, @mitsuhiko, was, of course, well aware of the existence and limitations of `pkg_resources` when he hacked this up.  I had just assumed there was a good reason as to why he didn't go that way.)

File: lektor/packages.py
Comment (dwt) [reply lektor/packages.py]: I've chosen this way of implementing this requires.txt -> requirements.txt cleanup step in a way that tries to be minimally invasive to the existing source code.

That being said, I too think that extracting the whole step including the `egg_info` generation would be much cleaner. As I now know that you're willing to take that change, I'll see what I can come up with.

**Comment by dwt** [comment]
I think that instead of trying to make the hack in a more estethically pleasing way, it would be better to fix the actual upstream bugs so that the whole installation can be reduced to `pip install --editable --target $folder $package` and get rid of the hack completely. Until then I don't quite see much value in investing a whole lot of time into this code.

File: lektor/packages.py
Comment (dwt) [reply lektor/packages.py]: @dairiki I've pushed a change that tries to get the spirit of what you said - what do you think about it?

File: lektor/packages.py
Comment (yagebu) [inline lektor/packages.py]: I think using a for-loop here is a slight abuse of notation.

File: tests/test_packages.py
Comment (yagebu) [inline tests/test_packages.py:33]: We're on Python 3.6 now, so passing these as Paths should work

File: lektor/packages.py
Comment (yagebu) [inline lektor/packages.py]: typo

**Review by yagebu — CHANGES_REQUESTED** [review]
I think it would be preferable if this code could do without the writing of `requirements.txt` - instead just read the requirements into a list and install that list of packages. That would make it easier to get rid of the for+yield hack and also make this part unit-testable.

**Comment by dairiki** [comment]
In working on #881, I've discovered that installing packages into Lektor's package cache is pretty slow.  This means that we should minimize the number package installs we do in the tests.  Running your four tests on my reasonably fast machine takes about 7 seconds.

After PR #881, which cleans up the slow `test_plugins.py`, the complete test suite — 304 tests in all — takes about 12 seconds to run.  (That doesn't include your new tests.) So the overhead to run your four new tests is significant.

This is just something to keep in mind.  Writing code to make it unit-testable, and then relying on unit testing to cover all the various edge cases can really speed up the test run time.

File: lektor/packages.py
Comment (dwt) [reply lektor/packages.py]: fixed

File: tests/test_packages.py
Comment (dwt) [reply tests/test_packages.py:33]: The original code works with string paths - so I didn't want to change that so as to not create a difference in how the code runs under tests vs. in real life. I can change it if it is important to you - but I wouldn't advise it.

File: lektor/packages.py
Comment (dwt) [reply lektor/packages.py]: That is true, but I do need a way to clean up the temporary files afterwards. I could use a context manager for that, but then the empty case (i.e. no requirements) gets more complicated and needs another in band signal to not call installation there.

Not quite sure what the best way forward here would be. I could pull out the whole tempdir into the surrounding function and handle it there, that would clean it up - but also pollute the outer function with a concern it doesn't really need to know about.

**Comment by dwt** [comment]
> I think it would be preferable if this code could do without the writing of `requirements.txt` - instead just read the requirements into a list and install that list of packages. That would make it easier to get rid of the for+yield hack and also make this part unit-testable.

I thought about that, and discarded it because I think that this would be a bad idea. Sorry, I should probably write this down as a comment in the source. The reason is this: pip needs to know about all the other packages that it is installing, so it can resolve dependencies correctly. If it doesn't know about the other packages, it can install incompatible versions, and especially versions that are different than what the plugin would install when it is installed un-editable, i.e. directly from pypi.

I think that is a hard no-go. That being said, I think the real push here should be to get the upstream bugs fixed, so that we can do `pip install --editable  $package_path --target $folder` and `pip install $package_name_from_pip --target $folder` to make the two cases the same.

Also, breaking the requirements down into individual items to be passed to pip on the command line requires more rewriting as requirements.txt and (to my knowledge probably requires.txt too) allow quite some additional syntax to specify constraints on the installed packages that the command line does not easily allow.

So… I think we need to stick with the requirements.txt, and / or go fix the upstream bugs (sooner).

**Comment by dwt** [comment]
> In working on #881, I've discovered that installing packages into Lektor's package cache is pretty slow. This means that we should minimize the number package installs we do in the tests. Running your four tests on my reasonably fast machine takes about 7 seconds.
> 
> After PR #881, which cleans up the slow `test_plugins.py`, the complete test suite — 304 tests in all — takes about 12 seconds to run. (That doesn't include your new tests.) So the overhead to run your four new tests is significant.
> 
> This is just something to keep in mind. Writing code to make it unit-testable, and then relying on unit testing to cover all the various edge cases can really speed up the test run time.

That's a good argument. I could rewrite the tests to not actually call the install, but just verify that it is called correctly. That being said, I would like to argue that the test-suite should have a layered approach. Some integration tests should test the whole roundtrip, to make sure that we get notified about interface changes, or changes in the way arguments are understood, while most of the test-suite then can stub out the actual installation.

Which leaves the question: where should the round-trips be, and perhaps should they marked in a special way so pytest executes them after the other tests, or even in it's own test run?

My guess would be that the `packages.py` module could be the right place to do the roundtrip tests, and then stub that module in other unit tests to speed them up.

What do you think about that?

**Comment by yagebu** [comment]
> The reason is this: pip needs to know about all the other packages that it is installing, so it can resolve dependencies correctly. If it doesn't know about the other packages, it can install incompatible versions, and especially versions that are different than what the plugin would install when it is installed un-editable, i.e. directly from pypi.

>Also, breaking the requirements down into individual items to be passed to pip on the command line requires more rewriting as requirements.txt and (to my knowledge probably requires.txt too) allow quite some additional syntax to specify constraints on the installed packages that the command line does not easily allow.

`pip` takes multiple on the command line just fine and also deals with PEP440 requirement specifiers, see https://pip.pypa.io/en/stable/reference/pip_install/#argument-handling, so I don't think these points are really problems with my suggested approach.

But maybe there's an even simpler approach that allows us to get rid of that whole egg-info parsing ugliness. Currently the approach is the following: 1) do an editable install the package without dependencies (because `--editable` and `--target` can't be combined) , 2) run setup.py egg_info 3) rewrite the requires from the egg-info and 4) install the dependencies from that. How about the following instead: 1) do a normal install of the package 2) do an editable install of the package without dependencies. That way, we install the package twice (so we trade one call to egg_info for an install of the plugin), but for a local plugin I'd not expect that shouldn't be a big performance hit. This would also allow this code to work for packages with an alternative build system.

Regarding the tests, I think it would be enough to have one test running a real pip install and then we can have the remainder just be unit tests (or some integration tests with the command calling stubbed out).

**Comment by dairiki** [comment]
> How about the following instead: 1) do a normal install of the package 2) do an editable install of the package without dependencies.

That would be ideal if it can be made to work.  I've just played around with it a bit, without luck.  The issue is that the second `--editable` install doesn't uninstall the normally installed version of the distribution.  Both the normal and --editable version end up installed (and, sadly, it appears it is the normal version that python loads when the package is imported.)  Manually uninstalling it is problematic, too, as `pip uninstall` doesn't appear to support a `--target` argument.

----

In other news, I noticed there is a [documented function](https://setuptools.readthedocs.io/en/latest/pkg_resources.html#parsing-utilities), `pkg_resources.split_sections` which can be used to parse `requires.txt`.
```
with open("requires.txt") as f:
    requirements = next(pkg_resources.split_sections(f))[1]
```
will work to get the list of requirements from just the first section (i.e. dropping the extras_require sections). (And it has the bonus of filtering out comments and blank lines.)

**Comment by dwt** [comment]
@dairiki While I do think that splitting out the requirements with `pkg_resources.split_sections()` is probably preferable to parsing the file myself, so I will change that.

However, I would really like to reflect on the fact, that this discussion feels quite a bit like bike-shedding.

There is a clear bug here, the pull request addresses it - and I think it would be worth merging that to get the bug fixed - and then starting the work on getting things even more better after that.

Regarding `pkg_resources.split_requirements()`: It is probably better to use that function, but it is also not super simple, for example, it will skip the default section if there are only `extras_require` and it needs still more extra hand holding if a package has no requirements.

**Comment by dwt** [comment]
I have now tried implementing the `split_sections()` way and have found that both it and the previous way of manually parsing the requires.txt file doesn't seem to be able to cope with more rare requires syntax, as setuptools seems to reformat a requirement of the form `watching_testrunner;python_version >= "3.6"` into this multi line monstrosity (of which I cannot find any documentation)

```ini

[:python_version >= "3.6"]
watching_testrunner
```
And which of course is neither compatible with pip requirements.txt as well as pip command line requirements.

Also going with handing in multiple  requirements to pip on the command line requires a new function, as the interface of the old one doesn't really suite itself to being extend to this.

So all in all I don't think this is worth it.

Also I would really like to stop the bike shedding and move on from this, so that I can finally get to the upstream libraries and fix the actual bug that causes all these workarounds.

**Comment by dwt** [comment]
Here's hope that https://github.com/pypa/pip/pull/9636 gets through soon and all of this becomes obsolete in a few years when users can use that new pip version.

**Comment by dwt** [comment]
Still, I would like to get this pull request through in some shape or form and not depend on the pip one, as no one knows how long that will take to propagate. So how do we continue here?

I would recommend to go back to a simple cleaning of requires.txt to generate requirements.txt.

Reasoning: AFAIK it is far more common to have `extra_requires` than environment markers, especially since setuptools deprecated test_requries, and suggested to use `extra_requires.test` as a replacement.

I think that this would make Lektor more usable until the pip pull request or an alternative gets through and we can get rid of the whole complication in lektor.

What do you guys think?

**Comment by dairiki** [comment]
@dwt What's your short-term motivator for this?  Is it just that lektor-atom is broken?

If so, what needs to be done to fix lektor-atom?  Does anyone know who can make a release?  With lektor-atom appearing to be part of the lektor project, and with lektor-atom currently breaking the CI tests for lektor-website, this is a bad look for the project and should be addressed.

Some immediate local fixes (which I'm sure you realize) are to create a private fork of lektor-atom, delete the extras_requires, and include that as a git submodule in the `packages` directory of your site.  Alternatively, you could (temporarily, until this is resolved) remove lektor-atom from your `.lektorproject`, install `lektor` and `lektor-atom@master` to a virtualenv, and run lektor from that.

Another alternative which might be worth considering is to disuse lektor-atom altogether.  It's really pretty straightforward to write a jinja template to construct a simple atom feed directly — no plugin necessary.  (I can share an example template, if you'd like.)

**Comment by dwt** [comment]
> @dwt What's your short-term motivator for this? Is it just that lektor-atom is broken?

Hm, maybe yes? I already have a local solution (the one you mentioned below) since quite some time. I would like to think that this is a bit of a more general problem as `test_requires` is deprecated and I really think that plugins ought to have a way to specify their testing-requirements in the setup.py file. I know you could just drop a requirements-test.txt in the project, but `extra_requires` are the recommended solution so I think we should support it until upstream fixes can get merged. (Which might take quite some time, as the discussion there shows)

> If so, what needs to be done to fix lektor-atom? Does anyone know who can make a release? With lektor-atom appearing to be part of the lektor project, and with lektor-atom currently breaking the CI tests for lektor-website, this is a bad look for the project and should be addressed.

That would be a good Idea regardless. If neccessary, I have worked on lektor-atom in the past and would be willing to take care of it, if given the neccessary rights. Not sure how you guys organise here, so maybe that's a no no?

All that being said - I think the pull request (maybe one or two commits back) still

- Improves the documentation dramatically why the code is structured the way it is
- Which bugs it depends upon to to get fixed before it can be removed
- Works a little bit better than before
- Without adding a lot of complexity.

I have to say, I think that this would be worth a merge.

**Comment by dairiki** [comment]
This is mostly off-topic, but...

> but extra_requires are the recommended solution

Who says?  There is debate about that. Myself, I don't think [tests] extras_requres are the way of the future.

The `setup.py`, these days, doesn't specify *how* to run tests, so it doesn't make much sense to me that that should be where test dependencies are specified.

Back in the old (`distutils`) days, there was `python setup.py test` which made it easy to run a specific `unittest` test suite. `Tests_require` specified the dependencies that should be installed in order to run that test suite. When that was in use, that made sense, since setup.py contained the complete specification as to how to run the python tests for the project. That is now [deprecated][], with the advice “Users looking for a generic test entry point independent of test runner are encouraged to use tox.”

Tox seems to currently be the de facto method of running python tests these days.  It supports the concept of multiple isolated test environments — each of which can require a different set of dependencies.  As an example of how this is useful, one can easily imagine that a Lektor plugin package might well want to run integration tests against multiple different versions of Lektor.  In that case, there is no one list of test requirements.

My preference is to put as much of the python test configuration as possible — that includes testing dependencies, as well as how to actually run the tests — in `tox.ini`.  (The `tox` configuration can alternatively be put into `pyproject.toml` or `setup.cfg` should you want to minimize the number of config files. Personally, I like having a `tox.ini`, as it serves as a fairly obvious marker that the project uses tox for testing.)

[deprecated]: <https://setuptools.readthedocs.io/en/latest/userguide/commands.html#test-build-package-and-run-a-unittest-suite>

Note that the [setuptools docs][], while they do make mention of the possibility of a `[tests]` or `[docs]` extra, do explicitly call out “two common use cases” for `extras_require`:

1. When a specific `console_scripts` entry point depends on extra dependencies.
2. To allow another distribution to require the extras in their own dependencies.

A `[test]` extras does not fall under either of those two use cases.  I view extras_require (and the rest of `setup.py`) as specifying packaging metadata, rather than development metadata.  A user of the distribution does not care what’s required to run the tests — most often the tests are not included in the (wheel) packaging, so the user can not run the tests anyway — so the test dependencies do really belong in package metadata.


[setuptools docs]: <https://setuptools.readthedocs.io/en/latest/userguide/dependency_management.html#optional-dependencies>

**Comment by dairiki** [comment]
As for the PR...

Keep in mind that I am not a project member and do not have merge/push rights, so this is just an opinion from the peanut gallery.

Also, keep in mind that, as already demonstrated, I am old and grumpy.  I don't mean to be overly critical.

Two points.

First, for some reason that I can't fully put into words, the idea of reading `requires.txt`, pulling the desired dependencies from that, writing them to `requirements.txt` and then passing that to pip sticks in my craw.  Writing of `requirements.txt` seems to be an extra step that serves no purpose.  I'd really rather see the requirements parsed from `requires.txt` and then those passed as command-line arguments to pip.

Personally, I would change the signature of `download_and_install_package` from
```python
def download_and_install_package(
    package_root, package=None, version=None, requirements_file=None
):
```
to
```python
def download_and_install_package(package_root, packages):
```
The only reason it currently supports passing `requirements_file` was to directly pass it the `requires.txt` from the egg-info.

Also moving the logic for forming the full requirement spec from `package` and `version` further up in the call chain will help if, at some time in the future, someone wants to put in the work to support more flexible ways of specifying version requirements in `.lektorproject`.  E.g. it would be nice to be able to do something like:
```ini
[packages]
lektor-plugin = >=1.0,<2
```

Second, the idea of adding code to "fix" something that mostly works, but only kind of by luck, with something that works a bit better, but still is known to be not fully correct or compatible does not give me warm fuzzy feelings.  Particularly when there are work-arounds which are (perhaps debately) not overly onerous.

**Comment by dwt** [comment]
Updated code according to the comments from @dairiki 

How are the @maintainers feeling about merging this? Is there still some more work required to make the test-suite faster? I somehow consider this something that can pretty much go completely away as soon as we can switch to `pip install --target $folder [--editable] $package_name_or_path_to_package`, as that would remove almost all of the workarounds. Till then, I kind of want the tests as a reminder that all the hackery a) still works and b) is a pain to maintain and should rather be removed ASAP.

What do you think?

File: lektor/packages.py
Comment (dwt) [reply lektor/packages.py]: For loop is gone in the current iteration.

File: lektor/packages.py
Comment (dairiki) [inline lektor/packages.py]: I'd suggest the name `requirements` (or perhaps `requirement_specifiers`), rather than `packages_with_versions` since that what pip [calls them](https://pip.pypa.io/en/stable/reference/pip_install/#requirement-specifiers).

File: lektor/packages.py
Comment (dairiki) [inline lektor/packages.py]: I know you just moved this code from where it used to be, but I'd rather see this as:
```python
requirement_spec = f"{package}=={version}" if version else package
download_and_install_packages(package_root, [requirement_spec])
```

(Perhaps a full `if` ... `else` ... to compute `requirement_spec` would be even more readable.)

File: lektor/packages.py
Comment (dairiki) [inline lektor/packages.py:181]: (Is there a good reason not in in-line this function within `requirements_from_unfinished_editable_install_at_path`?)

File: lektor/packages.py
Comment (dairiki) [inline lektor/packages.py]: (I wonder if there's a reason not to accumulate all of the non-local packages and install them in a single pip invocation, rather than installing them one at a time.)

**Review by dairiki — COMMENTED** [review]
FWIW, I like the current version much better.  Of course, I did have to leave a few comments...

Finally, given pypa/pip#9636 (great work!) is in the works, is it worth worrying about this at all, or would it be better to wait until we can just rely on pip to do the right thing?  (What's the chance Lektor will make a release before pip does?)

**Review by yagebu — COMMENTED** [review]
I'd prefer to not merge this, it still seems like a rather hackish stop-gap solution.

I believe that the underlying problem that we should address instead is our use of editable installs. I think they're too dangerous to do without explicit use action, since they pollute the python import path with whole packages directories. Their only advantage for Lektor is keeping track of changes, but that only works for code and not other packages changes, so I think doing normal install with some more explicit change tracking would be better. I'll open a (draft) PR removing our use of editable installs.

File: lektor/packages.py
Comment (dwt) [reply lektor/packages.py]: [x] done

File: lektor/packages.py
Comment (dwt) [reply lektor/packages.py]: I think a full if:else doesn't add much readability. Still I would really like to keep this patch somewhat minimal, given that it might become obsolete soon anyway.

File: lektor/packages.py
Comment (dwt) [reply lektor/packages.py:181]: I think the code reads and explains better what it does this way - the function is mostly self contained and doesn't need the context of the other function where it is used to be understood.

File: lektor/packages.py
Comment (dwt) [reply lektor/packages.py]: I would say that is something for a different pull request. This one should keep it's focus on fixing one thing first.

**Comment by dwt** [comment]
> FWIW, I like the current version much better. Of course, I did have to leave a few comments...

I hope those are addressed adequately in my commits and comments.

> Finally, given [pypa/pip#9636](https://github.com/pypa/pip/pull/9636) (great work!) is in the works, is it worth worrying about this at all, or would it be better to wait until we can just rely on pip to do the right thing? (What's the chance Lektor will make a release before pip does?)

I think that it is not quite as simple as waiting for a new pip release, as that would also mean that we have to add a dependency on that pip version and throw an error if the user has an older version installed.

I would do that in a separate (future) pull request, once pip is released and we had time to actually test it.

**Comment by dwt** [comment]
> I believe that the underlying problem that we should address instead is our use of editable installs. I think they're too dangerous to do without explicit use action, since they pollute the python import path with whole packages directories. Their only advantage for Lektor is keeping track of changes, but that only works for code and not other packages changes, so I think doing normal install with some more explicit change tracking would be better. I'll open a (draft) PR removing our use of editable installs.

Well, I don't think there is a good way to avoid editable installs of plugins you are working on completely. I don't know the history of why lektor did it's own editable installs, but I guess it could have something to do with the App GUI story that was present in the past? In that scenario, Lektor didn't really expose the virtualenv / place where the packages where exposed and needed to manage editable installs itself?

Some arguments that come to mind:
- Someone working on a plugin probably knows how to install Lektor himself into a virtualenv
- Someone distributing private plugins as submodules or in tree code as part of a site probably doesn't want collaborators to need to know how to install plugins himself (but for those cases a static install might be good enough?

As for something completely different: It might be possible to `pip install $plugin; pip uninstall $plugin --no-deps` to get jus the dependencies installed and then just add the `$LEKTOR_PROJECT/packages` folder to the python path?

**Comment by webknjaz** [comment]
Here's a wild thought: what if instead of requiring a plugin to be editable-installed, Lektor could just have a local plugin discovery mechanism that injects paths into `$PYTHONPATH`/`sys.path`?

This would make it importable and as for the deps, it's possible to monkey-patch `setuptools.setup`+`sys.argv` and invoke `setup.py` via something like `runpy.run_module('setup.py', name='__main__')`. Then the patched function would just collect the deps from the `install_requires` arg.

On the second thought, it could be easier to just require the plugin authors to use `setup.cfg` (which means requiring some minimum version of setuptools) — this would make parsing the data so much easier and would establish a better standard for describing the modules.

Another extreme would be relying on PEP517 to get the deps (there are hooks for this IIRC) but I think this doesn't allow one to extract the entrypoints as easily.

--------------------

@dairiki
> This is mostly off-topic, but...
>
> > but extra_requires are the recommended solution
>
> Who says? There is debate about that. Myself, I don't think [tests] extras_requres are the way of the future.

A part of the PyPA feels like having `tests` extras is a good idea. IIRC you can find a discussion about this on https://discuss.python.org. But then, no tools support installing extras-only, w/o the package itself. I'm in the camp of not using an extra for this too: test requirements should describe the env and therefore they should be strictly pinned unlike the open-ended runtime requirement specs, I find pip-tools to be a great solution for this problem.

**Comment by dairiki** [comment]
> I believe that the underlying problem that we should address instead is our use of editable installs. I think they're too dangerous to do without explicit use action, since they pollute the python import path with whole packages directories. 

Personally, since we're installing code from a single project's `packages` subdirectory into a private package library which is used only for that Lektor project, I don't see "pollution" as an issue.

The [docs make it clear](https://www.getlektor.com/docs/plugins/dev/) that some of the original motivation was to make it easy to work on the project-local plugins with changes to the code taking effect immediately (at least when running with `LEKTOR_DEV=1`.)

> Here's a wild thought: what if instead of requiring a plugin to be editable-installed, Lektor could just have a local plugin discovery mechanism that injects paths into $PYTHONPATH/sys.path?

> Another extreme would be relying on PEP517 to get the deps (there are hooks for this IIRC) but I think this doesn't allow one to extract the entrypoints as easily.

Without having thought about it too much, this seems like a great idea.  [Build a .dist-info](https://www.python.org/dev/peps/pep-0517/#prepare-metadata-for-build-wheel) in the source directory and add the directory to sys.path.
(That is all beyond the scope of this PR, though.)

**Comment by dwt** [comment]
@maintainers: I'd like to get this moving again.

Some points I would like to make:
- This pull request fixes a problem with existing released plugins
- It makes the code structure better, and more amenable to future expansion and fixes
- It also prepares to an eventual simplification when my fixes to pip are released
- It adds comments that explain why the code contains the hacks that it does
- It adds test cases for all these issues

I don't mind that much of this code gets removed in the long run, but I think it is an extremely potent base from which to start further refactorings from.

So I would like to get this merged and then continue working on further approaches. Because incremental improvement always wins.

**Comment by dairiki** [comment]
Closing in favor of #1065.

**Pull request closed without merging**
